lib/jack_tokenizer.rb

```diff
@@ -44,10 +57,10 @@ def identifier
   end
 
   def int_val
-    Integer(@current_token)
+    @current_token.to_i
```

Comment:
This switch from [`Kernel#Integer`](https://rubydoc.info/stdlib/core/Kernel:Integer) to [`String#to_i`](https://rubydoc.info/stdlib/core/String:to_i) was kind of accidental (you deleted the old implementation before adding the new one, and I think you just happened to pick a different method for doing the conversion) but it’s actually an improvement! Check it out:

```
>> '07'.to_i
=> 7
>> Integer('07')
=> 7

>> '08'.to_i
=> 8
>> Integer('08')
in `Integer': invalid value for Integer(): "08" (ArgumentError)
```

Can you explain what’s going on here? (I [mentioned](https://github.com/wildmaples/jack-compiler/pull/3#discussion_r738634603) this problem before but you didn’t take the bait.)

Reply:
Thanks for bringing it up again and forcing my eyes 👀 

I started with the `Kernel#Integer` documentation, in which this lines stand out: 
>  arg is a String, when base is omitted or equals zero, radix indicators (0, 0b, and 0x) are honored.

I know it has something to do with bases because for example, `0x` is the hexadecimal representation. So I googled: 

The fact that `Integer('07')` returns a 7 is misleading due to the assumption that the 7 is base-10. It's not returning a 7 in base-10, but in base-8 because a leading 0 indicates octal number system (source here). Hence when we try to provide `Integer` with the String argument `"08"` it returns the `ArgumentError` -- because it's not a valid input for that notation. 

This leads to the question, why is octal so important that we would have a notation dedicated to it? 

> Octal became widely used in computing when systems such as the UNIVAC 1050, PDP-8, ICL 1900 and IBM mainframes employed 6-bit, 12-bit, 24-bit or 36-bit words. Octal was an ideal abbreviation of binary for these machines because their word size is divisible by three (each octal digit represents three binary digits). So two, four, eight or twelve digits could concisely display an entire machine word. 

What I got from that is that it's the best numeric system to use for mainframes that want to use 6 to 36-bit words because those words can be fully represented in 8 digits.

Reply:
Perfect! 🎉🧑‍🎓

Although I’m sure you’ll agree the history lesson is fascinating in its own right, I guess the practical thing to remember is that although `Kernel#Integer` and `String#to_i` both take an optional `base` argument, they have different default behaviours when `base` isn’t provided. 

`String#to_i` defaults to decimal (base 10), which — let’s face it — is usually what you wanted. But `Kernel#Integer` tries to automatically _guess_ the base by looking at the contents of the string, so if you want to use `Kernel#Integer` and you _already know_ the string must contain a base 10 number (e.g. because it’s part of the source code of a Jack program, where only base 10 notation is supported), you should explicitly say so by passing the optional `base` argument:

```
>> Integer('08')
in `Integer': invalid value for Integer(): "08" (ArgumentError)
>> Integer('08', 10)
=> 8
```

Reply:
> a leading 0 indicates octal number system (source here)

cool story